core/primitives/src/errors.rs

```diff
@@ -27,73 +72,166 @@ impl std::fmt::Display for StorageError {
 impl std::error::Error for StorageError {}
 
 /// External
-#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "InvalidTx"]
 pub enum InvalidTxError {
-    InvalidSigner(AccountId),
-    SignerDoesNotExist(AccountId),
     InvalidAccessKey(InvalidAccessKeyError),
-    InvalidNonce(Nonce, Nonce),
-    InvalidReceiver(AccountId),
+    InvalidSigner {
+        signer_id: AccountId,
+    },
+    SignerDoesNotExist {
+        signer_id: AccountId,
+    },
+    InvalidNonce {
+        tx_nonce: Nonce,
+        ak_nonce: Nonce,
+    },
+    InvalidReceiver {
+        receiver_id: AccountId,
+    },
     InvalidSignature,
-    NotEnoughBalance(AccountId, Balance, Balance),
-    RentUnpaid(AccountId, Balance),
+    NotEnoughBalance {
+        signer_id: AccountId,
+        #[serde(with = "u128_dec_format")]
+        balance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
+    RentUnpaid {
+        signer_id: AccountId,
+        #[serde(with = "u128_dec_format")]
+        amount: Balance,
+    },
     CostOverflow,
     InvalidChain,
     Expired,
 }
 
-#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "InvalidAccessKey"]
 pub enum InvalidAccessKeyError {
-    AccessKeyNotFound(AccountId, PublicKey),
-    ReceiverMismatch(AccountId, AccountId),
-    MethodNameMismatch(String),
+    AccessKeyNotFound {
+        account_id: AccountId,
+        public_key: PublicKey,
+    },
+    ReceiverMismatch {
+        tx_receiver: AccountId,
+        ak_receiver: AccountId,
+    },
+    MethodNameMismatch {
+        method_name: String,
+    },
     ActionError,
-    NotEnoughAllowance(AccountId, PublicKey, Balance, Balance),
+    NotEnoughAllowance {
+        account_id: AccountId,
+        public_key: PublicKey,
+        #[serde(with = "u128_dec_format")]
+        allowance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
+}
+
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "Action"]
+pub struct ActionError {
```

Comment:
Please add comments.

Reply:
added

Reply:
Add a comment for this struct as well. Think about Rustdoc being generated and whether the field name is enough to understand why we need this.